### PR TITLE
feat: add function for winner to claim rewards

### PR DESCRIPTION
**Pull request by ftupas**
This PR is for issue https://github.com/onlydustxyz/starkonquest/issues/45.

The goal is to allow the winner to withdraw the tournament rewards by providing a `winner_withdraw` function. In this PR, I've done the following:
- create `winner_withdraw` **external** function that
  transfers ERC20 tokens from the tournament contract when called by the winner
- create `tournament_finished` and `rewards_withdrawn` **events**
- fix deprecated `stop_prank()` protostar cheatcode
image

- add tests for withdrawing rewards expecting reverts for withdrawals that are called
  -  when tournament is not yet finished
  - by zero address
  - anyone other than the winner
- update ITournament interface

### CODE REVIEW

File: contracts/tournament/library.cairo [hunk 1]
```diff
@@ -452,12 +472,46 @@ namespace tournament:
...
+
+        # Transfer rewards to winner
+        let (reward_amount) = reward_total_amount()
+        IERC20.transfer(
+            contract_address=reward_token_address, recipient=caller_address, amount=reward_amount
+        )
+        # Emit winner withdraw event
+        rewards_withdrawn.emit(caller_address, reward_amount)
+        return (TRUE)
+    end
```
Comment (ftupas) [inline contracts/tournament/library.cairo:514]: Entrypoint here. This functions contains the logic of how rewards can be withdrawn i.e. if the `caller_address` is the same as the winning `player_address` stored in the `tournament_winner_` storage variable

File: contracts/tournament/library.cairo [hunk 2]
```diff
@@ -626,11 +680,25 @@ namespace internal:
...
         if playing_ship_count == 1:
             # There is only one remaining ship, this is the winner of the tournament
             change_stage(tournament.STAGE_FINISHED)
+
+            # Record winner and emit event
+            let (winner_ship) = playing_ships_.read(0)
+            let (winner_address) = ship_player_.read(winner_ship)
+            let winner = Player(winner_address, winner_ship)
+            tournament_winner_.write(winner)
+            tournament_finished.emit(winner)
```
Comment (ftupas) [inline contracts/tournament/library.cairo:701]: Refactored the checking _if tournament is finished_ to this function. Also added logic to record the winner (player and ship addresses) as a `Player` struct into `tournament_winner_` storage variable and to emit it as an event marking the end of the tournament.

File: contracts/tournament/library.cairo [hunk 3]
```diff
@@ -35,6 +35,11 @@ end
 func tournament_name_() -> (res : felt):
 end
 
+# Winner of the tournament
+@storage_var
+func tournament_winner_() -> (player : Player):
+end
+
```
Comment (ftupas) [inline contracts/tournament/library.cairo:42]: Stores the tournament winner as a struct containing `player_address` and `ship_address`

File: contracts/tournament/test_tournament.cairo [hunk 4]
```diff
@@ -101,28 +102,28 @@ func test_close_registrations_with_good_ship_count{
     let (local context : TestContext) = test_internal.prepare(2, 2)
 
     # Start registrations
-    %{ start_prank(ids.context.signers.admin) %}
+    %{ stop_prank_admin = start_prank(ids.context.signers.admin) %}
     tournament.open_registrations()
     assert_that.stage_is(tournament.STAGE_REGISTRATIONS_OPEN)
-    %{ stop_prank() %}
+    %{ stop_prank_admin() %}
```
Comment (ftupas) [inline contracts/tournament/test_tournament.cairo:108]: Fix deprecated `stop_prank()` cheatcode

File: contracts/tournament/test_tournament.cairo [hunk 5]
```diff
@@ -475,7 +473,237 @@ func test_deposit_rewards_with_enough_allowance{
...
+        %{ stop_prank = start_prank(caller_address=2) %}
+        %{ expect_revert("TRANSACTION_FAILED", "Tournament: caller is not the final winner") %}
+        tournament.winner_withdraw()
+    end
+    %{ stop_prank() %}
+    return ()
+end
+
+@external
+func test_winner_withdraw{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}():
```
Comment (ftupas) [inline contracts/tournament/test_tournament.cairo:566]: Integration test of tournament and ERC20 token reward contract i.e. `ADMIN` deposits 100 tokens to the tournament contract, `PLAYER_1` and `PLAYER_2` register where after battling, the former wins and claims the rewards

**Review by ofux — APPROVED** [review]
Great contribution again! 👏

**Pull request merged**